src/lib.rs

```diff
@@ -141,14 +141,22 @@ impl Color {
         HSLA::from(self)
     }
 
-    /// Format the color as a HSL-representation string (`hsl(123, 50.3%, 80.1%)`).
+    /// Format the color as a HSL-representation string (`hsla(123, 50.3%, 80.1%, 0.4)`). If the
+    /// alpha channel is `1.0`, the simplified `hsl()` format will be used instead.
     pub fn to_hsl_string(&self, format: Format) -> String {
+        let space = if format == Format::Spaces { " " } else { "" };
         format!(
-            "hsl({:.0},{space}{:.1}%,{space}{:.1}%)",
-            self.hue.value(),
-            100.0 * self.saturation,
-            100.0 * self.lightness,
-            space = if format == Format::Spaces { " " } else { "" }
+            "hsl{a_prefix}({h:.0},{space}{s:.1}%,{space}{l:.1}%{a})",
+            a_prefix = if self.alpha == 1.0 { "" } else { "a" },
+            h = self.hue.value(),
+            s = 100.0 * self.saturation,
+            l = 100.0 * self.lightness,
+            space = space,
+            a = if self.alpha == 1.0 {
+                "".to_string()
+            } else {
+                format!(",{space}{:.3}", self.alpha, space = space)
+            }
```

Comment:
Maybe we could set
```rust
(a, a_prefix) = if self.alpha == 1.0 { … } else { … }
```
before the `format!` call to avoid having to duplicate the `self.alpha == 1.0` condition?

Concerning the `.3` digits for the alpha value: does that allow us to properly reconstruct all alpha values from 00 to ff? That is: can we properly serialize and deserialize i/255.0 for i = 0 … 255?

Reply:
I couldn't figure out a good solution for decimal/hex alpha values. Most hex alphas are irrational, so there's no good way to reliably convert back and forth between a decimal value. For example, hex `80` is `128 / 255`, or `0.501960784...` (non-terminating). I went with fixed precision here so I could at least write some form of test for the parser:

```rust
    // hex alpha doesn't line up nicely with decimal precision,
    // so just compare the debug output (3 digit precision)
    assert_eq!(
        format!("{:?}", Some(rgba(10, 0, 0, 0.502))),
        format!("{:?}", parse_color("0a000080"))
    );
```

but I'm not super happy with that either. I don't really have a good idea for how to allow alpha values to be converted freely between decimal and hex, or if there even is a good way to do that at all. Three digit precision seemed like a reasonable choice for human readability (personally I've only ever seen alphas out to 3 digits for very small values like `0.005`, and even that's rare, most alphas are 1-2 digit precision).

I'm open to ideas?

Reply:
> I couldn't figure out a good solution for decimal/hex alpha values. Most hex alphas are irrational, so there's no good way to reliably convert back and forth between a decimal value. For example, hex 80 is 128 / 255, or 0.501960784... (non-terminating).

Hm, yes. But that's not necessarily a problem, right? If alpha values are restricted to 8 bit anyway, then a few digits might be enough. What I meant was: 3 digits should be fine, because this succeeds (using Python for simplicity):
```python
for alpha_int in range(256):
    alpha_float = alpha_int / 255
    alpha_string_3digits = f"{alpha_float:.3f}"
    parsed_float = float(alpha_string_3digits)
    parsed_int = round(parsed_float * 255)

    assert alpha_int == parsed_int
```
… while two digits would not be enough.

Maybe we could write a unit test that does something similar?

Reply:
Good point, I guess I was getting too tied up with being able to do

```rust
assert_eq!(Some(rgba(10, 0, 0, 0.502)), parse_color("0a000080"));
```

instead of comparing `Debug` outputs. I think you're right that this isn't a real problem, so long as we can reliably go back and forth between hex 0..255 and three decimal precision. I'll translate your suggestion as a unit test 👍